### PR TITLE
Speedup spline rendering by >10x with large splines

### PR DESCRIPTION
**Pull request by veluca93**
This is achieved in multiple steps

- fix the distance cap formula
- avoid re-interpolating splines for every rect
- better SIMDfication

Performance could perhaps still be improved by changing the rendering method of splines in a more fundamental way.

Tested on a jxl-art file that makes heavy use of splines

```
Before:
1000 x 1000, geomean: 0.22 MP/s [0.22, 0.22], 5 reps, 0 threads.
  
No splines:
1000 x 1000, geomean: 14.05 MP/s [10.34, 14.16], 5 reps, 0 threads.
  
Fix distance formula:
1000 x 1000, geomean: 1.21 MP/s [1.17, 1.21], 5 reps, 0 threads.

Max butteraugli: 0.0039978079
3-norm: 0.001424
  
New DrawSegment:
1000 x 1000, geomean: 1.47 MP/s [1.42, 1.47], 5 reps, 0 threads.
                   
InitializeDrawCache:
1000 x 1000, geomean: 2.91 MP/s [2.76, 2.93], 5 reps, 0 threads.
```

### CODE REVIEW

**Comment by veluca93** [comment]
Using JXL's architecture diagram as a set of splines speed goes from 0.02 MP/s to ~0.95MP/s, for a 45x speedup, but this is still quite slow. I'll speed that one up too.

**Comment by veluca93** [comment]
> Using JXL's architecture diagram as a set of splines speed goes from 0.02 MP/s to ~0.95MP/s, for a 45x speedup, but this is still quite slow. I'll speed that one up too.

Improved this to ~15MP/s, for a total speedup of about 750x...

File: lib/jxl/splines.h [hunk 1]
```diff
@@ -79,6 +79,16 @@ class QuantizedSpline {
   int sigma_dct_[32] = {};
 };
 
+struct SplineSegment {
```
Comment (sboukortt) [inline lib/jxl/splines.h:82]: Roughly what is it that we call a segment here (for example why does it only have an x range)? Maybe a short explanatory comment wouldn’t hurt.

My understanding from the code is that it represents the intersection between a row in the image and the square of one Gaussian splat. Is that correct?

**Review by sboukortt — APPROVED** [review]

File: lib/jxl/splines.cc [hunk 2]
```diff
@@ -55,106 +55,115 @@ float ContinuousIDCT(const float dct[32], float t) {
...
 
-// Splats a single Gaussian on the image.
-void DrawGaussian(Image3F* const opsin, const Rect& opsin_rect,
-                  const Rect& image_rect, const Spline::Point& center,
-                  const float intensity, const float color[3],
-                  const float sigma, std::vector<int32_t>& xs,
-                  std::vector<int32_t>& ys,
-                  std::vector<float>& local_intensity_storage) {
+template <typename DF>
+void DrawSegment(DF, const SplineSegment& segment, bool add, size_t y, size_t x,
```
[the remark this answers is not in the document]
Comment (veluca93) [reply lib/jxl/splines.cc]: Absolutely and I don't know why I did otherwise :D

File: lib/jxl/splines.h (hunk 1 above)
Comment (veluca93) [reply lib/jxl/splines.h:86]: Yup, added a comment.

Comment (sboukortt) [reply lib/jxl/splines.h:86]: I think “line” -> “row” might be slightly less ambiguous. What do you think?

**Review by sboukortt — APPROVED** [review]

**Comment by sboukortt** [comment]
The size of one of the test images seems to have increased somewhat. How about a quick `optipng -o7 -strip all grayscale_patches_on_splines.png splines.png` to have some assurance that it is necessary?

**Comment by veluca93** [comment]
> The size of one of the test images seems to have increased somewhat. How about a quick `optipng -o7 -strip all blending/grayscale_patches_on_splines.png splines.png` to have some assurance that it is necessary?

Done, file sizes reduced.

**Review by sboukortt — APPROVED** [review]

**Comment by eustas** [comment]
```
FAILED: lib/CMakeFiles/jxl_dec-obj.dir/jxl/splines.cc.obj 
/usr/bin/clang++-7 --target=x86_64-w64-mingw32  -DJPEGXL_MAJOR_VERSION=0 -DJPEGXL_MINOR_VERSION=6 -DJPEGXL_PATCH_VERSION=0 -DJXL_INTERNAL_LIBRARY_BUILD -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" -I../ -I../lib/include -I../third_party/highway -I../third_party/brotli/c/include -Ilib/include -DHWY_BASELINE_TARGETS=HWY_SCALAR -ffunction-sections -fdata-sections -funwind-tables -Xclang -mrelax-all -Xclang -mconstructor-aliases -O3 -DNDEBUG -O2 -fvisibility=hidden -fvisibility-inlines-hidden   -Wno-builtin-macro-redefined -Wall -fmerge-all-constants -fno-builtin-fwrite -fno-builtin-fread -Wextra -Wc++11-compat -Warray-bounds -Wformat-security -Wimplicit-fallthrough -Wno-register -Wno-unused-function -Wno-unused-parameter -Wnon-virtual-dtor -Woverloaded-virtual -Wvla -Wc++2a-extensions -Wdeprecated-increment-bool -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wgnu-redeclared-enum -Winfinite-recursion -Wliteral-conversion -Wno-c++98-compat -Wno-unused-command-line-argument -Wprivate-header -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety-analysis -Wundefined-func-template -Wunreachable-code -Wunused-comparison -Wno-c++98-compat-pedantic -Wno-cast-align -Wno-double-promotion -Wno-float-equal -Wno-format-nonliteral -Wno-global-constructors -Wno-language-extension-token -Wno-missing-prototypes -Wno-shadow -Wno-shadow-field-in-constructor -Wno-sign-conversion -Wno-unused-member-function -Wno-unused-template -Wno-used-but-marked-unused -Wno-zero-as-null-pointer-constant -DJPEGXL_ENABLE_SKCMS=1 -DJPEGXL_BUNDLE_SKCMS=1 -std=c++11 -MD -MT lib/CMakeFiles/jxl_dec-obj.dir/jxl/splines.cc.obj -MF lib/CMakeFiles/jxl_dec-obj.dir/jxl/splines.cc.obj.d -o lib/CMakeFiles/jxl_dec-obj.dir/jxl/splines.cc.obj -c ../lib/jxl/splines.cc
fatal error: error in backend: Unable to copy EFLAGS physical register!
clang: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 7.0.0-3~ubuntu0.18.04.1 (tags/RELEASE_700/final)
Target: x86_64-w64-windows-gnu
Thread model: posix
InstalledDir: /usr/bin
clang: note: diagnostic msg: PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
```

**Comment by veluca93** [comment]
> ```
> FAILED: lib/CMakeFiles/jxl_dec-obj.dir/jxl/splines.cc.obj 
> /usr/bin/clang++-7 --target=x86_64-w64-mingw32  -DJPEGXL_MAJOR_VERSION=0 -DJPEGXL_MINOR_VERSION=6 -DJPEGXL_PATCH_VERSION=0 -DJXL_INTERNAL_LIBRARY_BUILD -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" -I../ -I../lib/include -I../third_party/highway -I../third_party/brotli/c/include -Ilib/include -DHWY_BASELINE_TARGETS=HWY_SCALAR -ffunction-sections -fdata-sections -funwind-tables -Xclang -mrelax-all -Xclang -mconstructor-aliases -O3 -DNDEBUG -O2 -fvisibility=hidden -fvisibility-inlines-hidden   -Wno-builtin-macro-redefined -Wall -fmerge-all-constants -fno-builtin-fwrite -fno-builtin-fread -Wextra -Wc++11-compat -Warray-bounds -Wformat-security -Wimplicit-fallthrough -Wno-register -Wno-unused-function -Wno-unused-parameter -Wnon-virtual-dtor -Woverloaded-virtual -Wvla -Wc++2a-extensions -Wdeprecated-increment-bool -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wgnu-redeclared-enum -Winfinite-recursion -Wliteral-conversion -Wno-c++98-compat -Wno-unused-command-line-argument -Wprivate-header -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety-analysis -Wundefined-func-template -Wunreachable-code -Wunused-comparison -Wno-c++98-compat-pedantic -Wno-cast-align -Wno-double-promotion -Wno-float-equal -Wno-format-nonliteral -Wno-global-constructors -Wno-language-extension-token -Wno-missing-prototypes -Wno-shadow -Wno-shadow-field-in-constructor -Wno-sign-conversion -Wno-unused-member-function -Wno-unused-template -Wno-used-but-marked-unused -Wno-zero-as-null-pointer-constant -DJPEGXL_ENABLE_SKCMS=1 -DJPEGXL_BUNDLE_SKCMS=1 -std=c++11 -MD -MT lib/CMakeFiles/jxl_dec-obj.dir/jxl/splines.cc.obj -MF lib/CMakeFiles/jxl_dec-obj.dir/jxl/splines.cc.obj.d -o lib/CMakeFiles/jxl_dec-obj.dir/jxl/splines.cc.obj -c ../lib/jxl/splines.cc
> fatal error: error in backend: Unable to copy EFLAGS physical register!
> clang: error: clang frontend command failed with exit code 70 (use -v to see invocation)
> clang version 7.0.0-3~ubuntu0.18.04.1 (tags/RELEASE_700/final)
> Target: x86_64-w64-windows-gnu
> Thread model: posix
> InstalledDir: /usr/bin
> clang: note: diagnostic msg: PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
> ```

I propose we fix this issue by getting rid of clang-7 - debian stable has >=8 now and even 18.04 has 8 in the main repos (interestingly, 7 is not in the main repos, but is in the -updates repos...)

**Comment by eustas** [comment]
(It is likely that this bug is finally fixed only in clang-9; it has been rediscovered multiple times under different conditions)

**Comment by eustas** [comment]
@deymo WDYT, should we ramp-up clang version, or we have clients who (cross-)compile JXL on clang-7?

**Comment by veluca93** [comment]
> (It is likely that this bug is finally fixed only in clang-9; it has been rediscovered multiple times under different conditions)

clang-9 is also in both ubuntu 18.04 and debian stable - for that matter, debian stable even has clang-11...

**Comment by veluca93** [comment]
> @deymo WDYT, should we ramp-up clang version, or we have clients who (cross-)compile JXL on clang-7?

IIRC we don't, at least for cross-compilation purposes.

**Pull request merged**